core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceConnection.java

```diff
@@ -544,9 +547,9 @@ public boolean addInferredStatement(Resource subj, IRI pred, Value obj, Resource
 			}
 			boolean modified = false;
 			if (contexts.length == 0) {
-				if (!hasStatement(explicitOnlyDataset, subj, pred, obj)) {
+				if (!hasStatement(explicitOnlyDataset, subj, pred, obj, NULL_CTX)) {
 					// only add inferred statements that aren't already explicit
-					if (!hasStatement(inferredOnlyDataset, subj, pred, obj)) {
+					if (!hasStatement(inferredOnlyDataset, subj, pred, obj, NULL_CTX)) {
```

Comment:
Just had a look at this after your remark. I'm not sure I follow the logic here to be honest. You've modified the check to only look in the null context. But what if the statement already existed (either inferred or explicit) in some other context? This adaptation will now report a new inferred statement even if in reality it might already have existed.

Reply:
It would seem more logical to me if this check explicitly looked at the context you're trying to insert the new statement to.

Reply:
> It would seem more logical to me if this check explicitly looked at the context you're trying to insert the new statement to.

It does. There are two branches. One for context.length == 0 and one for length == 1. This is for length == 0.

Reply:
Ah, right, I missed that in the diff!